lib/Dialect/Seq/Transforms/SinkClockGates.cpp

```diff
@@ -0,0 +1,260 @@
+//===- SinkClockGates.cpp - Sink clock gates to its users. -----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Dialect/HW/HWInstanceGraph.h"
+#include "circt/Dialect/HW/HWOpInterfaces.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+#include "circt/Dialect/Seq/SeqPasses.h"
+#include "circt/Dialect/Seq/SeqTypes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/MLIRContext.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/Debug.h"
+#include <string>
+
+namespace circt {
+namespace seq {
+#define GEN_PASS_DEF_SINKCLOCKGATES
+#include "circt/Dialect/Seq/SeqPasses.h.inc"
+} // namespace seq
+} // namespace circt
+
+#define DEBUG_TYPE "sink-clock-gates"
+
+using namespace circt;
+using namespace seq;
+using namespace hw;
+
+namespace {
+struct SinkClockGatesPass
+    : public circt::seq::impl::SinkClockGatesBase<SinkClockGatesPass> {
+  using SinkClockGatesBase::SinkClockGatesBase;
+
+  // TODO: Can be moved to some utils !!
+  /// Append inputs to the instance, and also update the referenced modules.
+  void appendInputsToInstance(HWInstanceLike instance,
+                              ArrayRef<std::pair<StringAttr, Value>> inputs,
+                              ArrayRef<HWModuleOp> refModules) {
+
+    SmallVector<Attribute> argNames;
+    if (auto args = instance->getAttrOfType<ArrayAttr>("argNames"))
+      argNames.insert(argNames.begin(), args.begin(), args.end());
+
+    SmallVector<Value> newOperands;
+    newOperands.reserve(inputs.size());
+    SmallVector<std::tuple<StringAttr, Type, Location>> newInputs;
+    newInputs.reserve(inputs.size());
+    for (const auto &[name, value] : inputs) {
+      // If no argNames attribute, donot add it.
+      if (!argNames.empty())
+        argNames.push_back(name);
+      // Additional operands to the Instance op.
+      newOperands.push_back(value);
+      // Prepare the additional input ports that need to be added to the
+      // referenced module.
+      newInputs.emplace_back(name, value.getType(), value.getLoc());
+    }
+    // Add the new operands to the instance.
+    instance->insertOperands(instance->getNumOperands(), newOperands);
+    if (!argNames.empty())
+      instance->setAttr("argNames",
+                        ArrayAttr::get(instance->getContext(), argNames));
+    // Update all the modules with the new input ports.
+    for (auto refMod : refModules) {
+      refMod.appendPorts(newInputs, {});
+    }
+  }
```

Comment:
This is a very handy function! Should this potentially accept multiple (instance, inputs) tuples, since a module might be instantiated multiple times and you'd need to update each of the instances in turn?

Reply:
You are right, fixed it.